### PR TITLE
[nnc] Fix half2float conversion and re-enable float16

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64196
* __->__ #64147

Fused float16 operations are cast to float for arithmetic, and then cast back down to float16 for storage, since (1) fused kernels tend to be bandwith bound anyways, and (2) some older gpus had pretty bad fp16 perf, and (3) this change actually improves numeric accuracy.

The problem for NNC was that we weren't properly converting all the intermediate ops; we were just leaving them all as "half" and hoping that (because the CUDA C++ we spit out doesn't usually name the types explicitly) that it would work out OK.  It sort of does, but weird bugs crop up, like the final cast back to half (`result[...] = half(...expr...);` would get "optimized" away, because the IR already thought the result was half.

This change properly converts all the float16 arithmetic in a kernel to float32 in the IR, getting rid of this class of weird bugs.

### CODE REVIEW

**Comment by bertmaher** [comment]
We disabled half in the fuser because of issues like #61382.  This PR fixes some remaining issues in half-rewriting and re-enables fp16 in the fuser.

File: test/test_jit_fuser_te.py [hunk 1]
```diff
@@ -95,7 +96,7 @@ def setUp(self):
         ]
         self.fp_dtypes = [
             # TODO: Add back when https://github.com/pytorch/pytorch/issues/55905 is closed
```
Comment (ZolotukhinM) [inline test/test_jit_fuser_te.py:98]: Nit: remove TODO

File: test/test_jit_fuser_te.py [hunk 2]
```diff
@@ -1768,7 +1790,9 @@ def test_type_as_cat(self):
         with inline_fusion_groups():
             def eager(x, y):
                 return torch.cat((x, y.type_as(x)), dim=1)
-            for dtype1, dtype2 in product(self.dtypes, self.dtypes):
+            dtypes = self.dtypes.copy()
+            dtypes.remove(torch.float16)
```
Comment (ZolotukhinM) [inline test/test_jit_fuser_te.py:1794]: Nit: might be worth of a comment?

File: test/test_jit_fuser_te.py [hunk 3]
```diff
@@ -1936,6 +1960,22 @@ def bn_neither(i, x):
             for fn in [bn, bn_no_weight, bn_no_bias, bn_neither]:
                 test(fn, (i, x))
 
+    @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA for fp16")
+    def test_bn_half_61382(self):
```
Comment (ZolotukhinM) [inline test/test_jit_fuser_te.py:1964]: We already have an equivalent test, it's probably better to just re-enable it:
https://github.com/pytorch/pytorch/blob/44e3ed88c9a1bd9ee6b0168ba5271a2c6b006cc8/test/test_tensorexpr.py#L1225-L1241

File: torch/csrc/jit/tensorexpr/ir_verifier.cpp [hunk 4]
```diff
@@ -119,7 +119,19 @@ void IRVerifier::visit(IfThenElsePtr v) {
...
+    }
+    IRVisitor::visit(v);
+    return;
+  }
   // TODO: add a check for OpArgCount and op_type
+  for (auto const& param : v->params()) {
+    if (param->dtype() != v->dtype()) {
+      throw malformed_ir("bad dtype in intrinsic arg");
+    }
+  }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_verifier.cpp:134]: Is it a requirement that all parameters should have the same dtype as the result value? I guess it's true at the moment, but I wonder if it's appropriate to enforce this condition in IR verifier

File: torch/csrc/jit/tensorexpr/half_support.h [hunk 5]
```diff
@@ -128,6 +128,46 @@ class HalfRewriter : public IRMutator {
...
+  template <typename T>
+  ExprPtr mutateArithmetic(T v) {
+    IRMutator::mutate(v);
+    if (v->dtype().scalar_type() == c10::kHalf) {
+      v->set_dtype(v->dtype().cloneWithScalarType(c10::kFloat));
+    }
+    return v;
+  }
+
+  ExprPtr mutate(AddPtr v) override {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/half_support.h:140]: Do you mind adding a brief description of what was broken in that test and how this change fixes it?

**Review by ZolotukhinM — APPROVED** [review]
Looks good modulo minor comments!

File: torch/csrc/jit/tensorexpr/ir_verifier.cpp (hunk 4 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/ir_verifier.cpp:134]: Yeah it's definitely not universally true (as you see I had to special-case `isnan`), but it's otherwise true for the set of intrinsics we have.  I think llvm would complain anywhere this doesn't hold, since the type signatures wouldn't match.

File: torch/csrc/jit/tensorexpr/half_support.h (hunk 5 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/half_support.h:140]: Sure thing, will update the summary.
